pagespeed/envoy/envoy_url_async_fetcher_test.cc

```diff
@@ -0,0 +1,215 @@
+        fetcher_timeout_ms_(FetcherTimeoutMs()) {}
+
+  virtual void SetUp() { SetUpWithProxy(""); }
+
+  static int64 FetcherTimeoutMs() {
+    return RunningOnValgrind() ? kFetcherTimeoutValgrindMs : kFetcherTimeoutMs;
+  }
+
+  void SetUpWithProxy(const char* proxy) {
+    const char* env_host = getenv("PAGESPEED_TEST_HOST");
+    if (env_host != NULL) {
+      test_host_ = env_host;
+    }
+    if (test_host_.empty()) {
+      test_host_ = kFetchHost;
+    }
+    GoogleString fetch_test_domain = StrCat("//", test_host_);
+    timer_.reset(Platform::CreateTimer());
+    statistics_.reset(new SimpleStats(thread_system_.get()));
+    EnvoyUrlAsyncFetcher::InitStats(statistics_.get());
+    envoy_url_async_fetcher_.reset(
+        new EnvoyUrlAsyncFetcher(proxy, thread_system_.get(), statistics_.get(), timer_.get(),
+                                 fetcher_timeout_ms_, &message_handler_));
+    mutex_.reset(thread_system_->NewMutex());
+
+    // Set initial timestamp so we don't roll-over monitoring stats right after
+    // start.
+    statistics_->GetUpDownCounter(EnvoyStats::kEnvoyFetchLastCheckTimestampMs)
+        ->Set(timer_->NowMs());
+  }
+
+  virtual void TearDown() {
+    // Need to free the fetcher before destroy the pool.
+    delete envoy_fetch_;
+    envoy_url_async_fetcher_.reset(NULL);
+    timer_.reset(NULL);
+    STLDeleteElements(&fetches_);
+  }
+
+  // Adds a new URL & expected response to the url/response structure, returning
+  // its index so it can be passed to StartFetch/StartFetches etc.
+  int AddTestUrl(const GoogleString& url, const GoogleString& content_start) {
+    urls_.push_back(url);
+    content_starts_.push_back(content_start);
+    int index = fetches_.size();
+    fetches_.push_back(new EnvoyTestFetch(
+        RequestContext::NewTestRequestContext(thread_system_.get()), mutex_.get()));
+    return index;
+  }
+
+  void StartFetch(int idx) {
+    fetches_[idx]->Reset();
+    envoy_url_async_fetcher_->Fetch(urls_[idx], &message_handler_, fetches_[idx]);
+  }
+
+  void StartFetches(size_t first, size_t last) {
+    for (size_t idx = first; idx <= last; ++idx) {
+      StartFetch(idx);
+    }
+  }
+
+  std::vector<EnvoyTestFetch*> fetches_;
+  std::vector<GoogleString> content_starts_;
+  std::vector<GoogleString> urls_;
+
+  scoped_ptr<EnvoyUrlAsyncFetcher> envoy_url_async_fetcher_;
+  scoped_ptr<Timer> timer_;
+  scoped_ptr<ThreadSystem> thread_system_;
+  scoped_ptr<AbstractMutex> mutex_;
+  scoped_ptr<SimpleStats> statistics_;
+
+  int64 flaky_retries_;
+  int64 fetcher_timeout_ms_;
+
+  EnvoyTestFetch* envoy_fetch_;
+  GoogleString test_host_;
+  MockMessageHandler message_handler_;
+};
+
+TEST_F(EnvoyUrlAsyncFetcherTest, FetchURL) {
+  envoy_fetch_ = new EnvoyTestFetch(
+            RequestContext::NewTestRequestContext(thread_system_.get()),
+            mutex_.get());
+  envoy_url_async_fetcher_->Fetch("http://localhost:80", &message_handler_, envoy_fetch_);
```

Comment:
Hmm, this isn't going to work in CI, right? (I assume there's no server listening there).
Maybe use on of the test urls that Serf uses targetting modpagespeed.com?

Reply:
test case updated with fetching of url **http://selfsigned.modpagespeed.com/mod_pagespeed_example/index.html**
Code pushed in https://github.com/apache/incubator-pagespeed-mod/pull/1969/commits/8a92117ec74a99b4a1dffff479a71ca1b5db7717